### PR TITLE
feat(procedure): auto split large value to multiple values

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

auto split large value into multiple values

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/common/meta/src/state_store.rs [hunk 1]
```diff
@@ -64,20 +70,75 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
...
-            .put(PutRequest {
-                key: with_prefix(key).into_bytes(),
-                value,
-                ..Default::default()
-            })
-            .await
-            .map_err(BoxedError::new)
-            .context(PutStateSnafu { key })?;
-        Ok(())
+        let split = split_value(&value, self.max_size_per_value);
```
Comment (fengjiachun) [inline src/common/meta/src/state_store.rs:93]: Better to move the value into `split_value`

File: src/common/meta/src/state_store.rs [hunk 2]
```diff
@@ -64,20 +70,75 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
...
+    Single(&'a [u8]),
+    Multiple(Vec<&'a [u8]>),
+}
+
+fn split_value(value: &[u8], max_size_per_value: Option<usize>) -> SplitValue<'_> {
+    if let Some(max_size_per_value) = max_size_per_value {
+        if value.len() < max_size_per_value {
+            SplitValue::Single(value)
+        } else {
+            SplitValue::Multiple(value.chunks(max_size_per_value).collect::<Vec<_>>())
```
Comment (fengjiachun) [inline src/common/meta/src/state_store.rs:83]: `itertools::Itertools` can chunks and avoid cloning

File: src/common/procedure/src/store.rs [hunk 3]
```diff
@@ -143,16 +144,17 @@ impl ProcedureStore {
         // 8 should be enough for most procedures.
         let mut step_keys = Vec::with_capacity(8);
         let mut finish_keys = Vec::new();
-        while let Some((key, _)) = key_values.try_next().await? {
-            let Some(curr_key) = ParsedKey::parse_str(&self.proc_path, &key) else {
+        while let Some((keyset, _)) = key_values.try_next().await? {
```
Comment (fengjiachun) [inline src/common/procedure/src/store.rs:147]:
```suggestion
        while let Some((key_set, _)) = key_values.try_next().await? {
```

File: src/common/meta/src/state_store.rs [hunk 4]
```diff
@@ -64,20 +70,75 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
...
 }
 
+enum SplitValue<'a> {
+    Single(&'a [u8]),
+    Multiple(Vec<&'a [u8]>),
+}
+
+fn split_value(value: &[u8], max_size_per_value: Option<usize>) -> SplitValue<'_> {
+    if let Some(max_size_per_value) = max_size_per_value {
+        if value.len() < max_size_per_value {
```
Comment (fengjiachun) [inline src/common/meta/src/state_store.rs:80]:
```suggestion
        if value.len() <= max_size_per_value {
```

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/state_store.rs [hunk 5]
```diff
@@ -64,20 +70,80 @@ fn decode_kv(kv: KeyValue) -> Result<(String, Vec<u8>)> {
...
+                let operations = values
+                    .into_iter()
+                    .enumerate()
+                    .map(|(idx, value)| {
+                        let key = if idx > 0 {
+                            KeySet::with_segment_suffix(&key, idx)
+                        } else {
+                            key.to_string()
+                        };
+                        let kv_backend = self.kv_backend.clone();
```
Comment (evenyag) [inline src/common/meta/src/state_store.rs:126]: question: Does kv backend provide a batch put method?

File: src/common/procedure/src/store/util.rs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                    parsed_segments
+                        .into_iter()
+                        .map(|(key, _)| key)
+                        .collect::<Vec<_>>()
+                )
+            }
+        );
+
+        let segment_values = parsed_segments.into_iter().map(|(_, value)| value);
+        let mut values = vec![value];
```
Comment (evenyag) [inline src/common/procedure/src/store/util.rs:79]: Use `with_capacity`?

File: src/common/procedure/src/store/util.rs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    Collecting,
+    End,
+}
+
+pub type Collector = dyn Fn(CollectingState) -> Result<(KeySet, Vec<u8>)> + Send;
+pub type Upstream = dyn Stream<Item = Result<(String, Vec<u8>)>> + Send;
+
+/// A stream collects multiple values into a single key-value pair.
+pub struct MultipleValuesStream {
+    upstream: Pin<Box<Upstream>>,
```
Comment (evenyag) [inline src/common/procedure/src/store/util.rs:114]: Is it possible that the output ordering of `Upstream` is unspecific?

File: src/common/procedure/src/store/util.rs [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+            collecting: None,
+            collector,
+        }
+    }
+}
+
+impl Stream for MultipleValuesStream {
+    type Item = Result<(KeySet, Vec<u8>)>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```
Comment (evenyag) [inline src/common/procedure/src/store/util.rs:134]: I'd prefer to use async_stream for readability.

File: src/common/procedure/src/store/util.rs [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+#[derive(Debug)]
+enum MultipleValuesStreamState {
+    Idle,
+    Collecting,
+    End,
+}
+
+pub type Collector = dyn Fn(CollectingState) -> Result<(KeySet, Vec<u8>)> + Send;
```
Comment (evenyag) [inline src/common/procedure/src/store/util.rs:109]: This might be overdesign......

File: src/common/meta/src/state_store.rs (hunk 5 above)
Comment (WenyXu) [reply src/common/meta/src/state_store.rs:126]: Yes, but the batch-put method will reach the size limit. 

https://etcd.io/docs/v3.5/dev-guide/limit/
> By default, the maximum size of any request is 1.5 MiB.

File: src/common/procedure/src/store/util.rs (hunk 7 above)
Comment (WenyXu) [reply src/common/procedure/src/store/util.rs:113]: No, it requires output ordering of `Upstream` in the lexical order.

File: src/common/procedure/src/store/util.rs (hunk 9 above)
Comment (WenyXu) [reply src/common/procedure/src/store/util.rs]: Removed

File: src/common/procedure/src/store/util.rs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+                    } else {
+                        // Starts to collect next key value pair.
+                        collecting = Some(CollectingState::new(key, value));
+                        yield multiple_values_collector(current)?;
+                    }
+                }
+                None => collecting = Some(CollectingState::new(key, value)),
+            }
+        }
+        if let Some(current) = collecting.take(){
```
Comment (evenyag) [inline src/common/procedure/src/store/util.rs:126]:
```suggestion
        if let Some(current) = collecting.take() {
```

**Review by evenyag — APPROVED** [review]
Rest LGTM

**Pull request merged**